numpy/core/shape_base.py

```diff
@@ -365,78 +365,91 @@ def stack(arrays, axis=0, out=None):
-                     ``x`` itself
-        * ``entering``: bool. The result of ``recurse_if(value)``
-        """
-        do_recurse = self.recurse_if(x)
-        yield index, x, do_recurse
-
-        if not do_recurse:
-            return
-        for i, xi in enumerate(x):
-            # yield from ...
-            for v in self.walk(xi, index + (i,)):
-                yield v
+    def format_index(index):
+        idx_str = ''.join('[{}]'.format(i) for i in index if i is not None)
+        return 'arrays' + idx_str
+    if type(arrays) is tuple:
+        # not strictly necessary, but saves us from:
+        #  - more than one way to do things - no point treating tuples like
+        #    lists
+        #  - horribly confusing behaviour that results when tuples are
+        #    treated like ndarray
+        raise TypeError(
+            '{} is a tuple. '
+            'Only lists can be used to arrange blocks, and np.block does '
+            'not allow implicit conversion from tuple to ndarray.'.format(
+                format_index(parent_index)
+            )
+        )
+    elif type(arrays) is list and len(arrays) > 0:
+        indexes, arr_ndims = zip(*[_block_check_depths_match(arr, parent_index + [i])
+                                   for i, arr in enumerate(arrays)])
+
+        first_index = indexes[0]
+        for i, index in enumerate(indexes):
+            if len(index) != len(first_index):
+                raise ValueError(
+                    "List depths are mismatched. First element was at depth "
+                    "{}, but there is an element at depth {} ({})".format(
+                        len(first_index),
+                        len(index),
+                        format_index(index)
+                    )
+                )
+        return first_index, max(arr_ndims)
+    elif type(arrays) is list and len(arrays) == 0:
+        # We've 'bottomed out' on an empty list
+        return parent_index + [None], 0
+    else:
+        # We've 'bottomed out' - arrays is either a scalar or an array
+        return parent_index, _nx.ndim(arrays)
+
+
+def _block(arrays, max_depth, result_ndim):
+    """
+    Internal implementation of block. `arrays` is the argument passed to
+    block. `max_depth` is the depth of nested lists within `arrays` and
+    `result_ndim` is the greatest of the dimensions of the arrays in
+    `arrays` and the depth of the lists in `arrays` (see block docstring
+    for details).
+    """
+    def atleast_nd(a, ndim):
+        # Ensures `a` has at least `ndim` dimensions by prepending
+        # ones to `a.shape` as necessary
+        return array(a, ndmin=ndim, copy=False, subok=True)
+
+    def block_recursion(arrays, depth=0):
+        if depth < max_depth:
+            if len(arrays) == 0:
+                raise ValueError('Lists cannot be empty')
+            arrs = [block_recursion(arr, depth+1) for arr in arrays]
+            return _nx.concatenate(arrs, axis=-(max_depth-depth))
```

Comment:
Wouldn't `-(max_depth-depth)` just be `(depth-max_depth)`?

Reply:
Also slight personal preference to having spaces around binary operators for readability.

Reply:
>Wouldn't `-(max_depth-depth)` just be `(depth-max_depth)`?

It would, yeah. The reason I wrote it that way round was because I wanted to match the docstring as closely as possible (my original motivation for this pr was to make it a bit clearer to people reading the code how block actually worked). The docstring says:
>Blocks in the innermost lists are concatenated (see `concatenate`) along the last dimension (-1), then these are concatenated along the second-last dimension (-2), and so on until the outermost list is reached.

To me the correspondence between the docstring and the code is ever so slightly more clear with the expression the way round that it currently is, and I think the effect on performance is probably negligable.

Reply:
>Also slight personal preference to having spaces around binary operators for readability.

I often instinctively shorten things when they're being fed to a keyword argument, because you don't normally put spaces around the `=` sign. Don't
```python
axis=-(max_depth - depth)
```
and
```python
axis=depth - max_depth
```
look a bit weird?

Reply:
The style point is minor. Though there are many tools (e.g. `flake8`) that consider not having spaces around binary operators an error regardless of context. Normally this is argued from the standpoint of readability (including in PEP8 when it was introduced). Hence why I mentioned it. Adding parentheses around the keyword argument's value help guide the eye when following these style tools' recommendations. Not strongly attached to this styling nit though.